### ISSUE TITLE
[Tracking] New pull request assignment proposal

### ISSUE THREAD

**Issue by apiraino** [rust-lang/triagebot#1753]
This issue tracks the progress of the implementation of a new workflow for assigning pull requests to the Rust project contributors.

**2023-12-12 UPDATE**: After another meeting with T-infra, we agreed a second plan detailing how to split the work further. The new plan is detailed below.

**2023-11-08 UPDATE:** After a meeting with T-infra, it was remarked that the first implementation (#1719) was hard to review so we agreed a plan to split into smaller bites (#1745)

- [ ] Step 1
  - add the DB table, strip down DB fields to user_id,assigned_prs
  - add code to keep track of assign/unassign'ed PRs
  - no migration/import/sync tooling
- [ ] Step 2
  - add logic to assign PRs based on a `count()` of your assigned PRs
  - prepare ENV VAR to switch upon restarting triagebot
- [ ] Step 3 (to be clarified)
  - HTTP endpoint to sync assignment counts with GitHub
- [ ] Step 4 (to be clarified)
  - Tackle the preferences: HTML or chat-based?
- [ ] Step 5 (to be clarified)
  - a tool (cli?) for migration/sync of PRs from Github?
  - how to sync PRs when/if bors webooks fail?
  - backoffice: add a button to the contributor to resync their own situation

---

## Summary

The current pull request assignment is basically randomly assigned among team members. The new proposed workflow has the following features:
- each Rust project contributor belonging to a team (I'd start will a small cohort of testers) can set their own review capacity and time off using a web backoffice
- the web backoffice will persiste these data in a DB
- everytime a pull request assignment is invoked with `r? <team> or <team_member>` the triagebot will check the current availability and workload of the candidates and assign the pull request to the team member less busy, excluding automatically those that are on time off in that moment.

A slightly old design document is at [this HackMD link](https://hackmd.io/GqXA4nyTQ4m__hRZUayVRg).

## Permissions and Authorization

This backoffice could be thought as a simple form listing all team members. It is served by the triagebot (from triagebot.infra.rust-lang.org). Access to this backoffice is gated by a GitHub App. Users allowed to this backoffice:
1) must be logged into GitHub
2) must agree to the share some details with a GitHub App (only to read their profile)
3) must be a Rust project team member
4) the team they belong must be whitelisted in the env var `NEW_PR_ASSIGNMENT_TEAMS`

Team leaders are considered administrators and can see the preferences of every team member. Normal team members can only see their own preferences and those of team members that agree to share theirs within the team.

## DEPLOYMENT

There will a few env vars to enable the new PR assignment and retrict access to this backoffice.

The idea would be to iterate on this pull request and iron out all the wrinkles visible without deployment:
- Deploy this changes but keep the new PR assignment DISABLED. The web backoffice will still be accessible.
- Team members can access that, start playing with with it fill their preferences
- When we are all confident about these changes, we can flip the env variable `USE_NEW_PR_ASSIGNMENT` and have the team listed in `NEW_PR_ASSIGNMENT_TEAMS` use it for real.
- After a first period of tests, add more teams
